clients/drcachesim/analyzer.cpp

```diff
@@ -373,6 +415,32 @@ analyzer_tmpl_t<RecordType, ReaderType>::init_scheduler_common(
     return true;
 }
 
+template <typename RecordType, typename ReaderType>
+void
+analyzer_tmpl_t<RecordType, ReaderType>::add_noise_generator_to_workloads(
+    std::vector<typename sched_type_t::input_workload_t> &workloads,
+    std::unique_ptr<noise_generator_info_t> info)
+{
+    // Add noise generator reader to workloads.
+    for (uint64_t pid_idx = 0; pid_idx < info->num_processes; ++pid_idx) {
+        // Assumptions in the scheduler disallow pid to be 0.
+        uint64_t pid = pid_idx + 1;
+        std::vector<typename sched_type_t::input_reader_t> readers;
+        for (uint64_t tid_idx = 0; tid_idx < info->num_threads_per_process; ++tid_idx) {
+            // Assumptions in the scheduler disallow tid to be 0.
+            uint64_t tid = tid_idx + 1;
+            auto noise_generator =
+                get_noise_generator(static_cast<addr_t>(pid), static_cast<addr_t>(tid),
+                                    info->num_records_to_generate);
+            auto noise_generator_end = get_noise_generator_end();
+            readers.emplace_back(std::move(noise_generator),
+                                 std::move(noise_generator_end),
+                                 static_cast<memref_tid_t>(tid));
+        }
+        workloads.emplace_back(std::move(readers));
```

Comment:
I think the noise generator module should provide convenience routines like this, not just for the analyzer but for simulators who otherwise would have to duplicate this logic including figuring out what the pid and tid should be which does not seem good.  I.e., I think this code should live inside the noise generator.

Reply:
A noise generator is a `reader_t` itself, IMHO it would look weird to have a `noise_Generator_t` add itself to the scheduler's vector of workloads.
Moved this "utility routine" into `noise_generator_factory_t`, so we can reuse it.